Allow optionally turning off customized file manager

#181 uses a VFS approach to protect `.class` files generation inside a transaction. This in theory should work for all files generated by javac, however, while it works for `.class` files, which is the majority case, `resource` (could be projects using annotation processor) and `.java` (An example is jmh that first generates `.java` files, then compiles them into `.class` files) files currently fail under this approach, see stack trace in [1].

The root cause is a current javac limitation, [somewhere in `JavacFileManager`](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b27/com/sun/tools/javac/file/JavacFileManager.java#638) there is a hard coded `instanceof` check, so when `FileObject` or `JavaFileObject` are wrapped into `ForwardingFileObject`/`ForwardingJavaFileObject`, instanceof check would fail because the check is on the forwarding object, not its wrapped object.

```
public boolean isSameFile(FileObject a, FileObject b) {
       nullCheck(a);
       nullCheck(b);
       if (!(a instanceof BaseFileObject))
           throw new IllegalArgumentException("Not supported: " + a);
       if (!(b instanceof BaseFileObject))
           throw new IllegalArgumentException("Not supported: " + b);
       return a.equals(b);
   }
```

Assume upstream change (I opened a request internally at Twitter) will take some time, we need some workaround to turn off the customized manager if project uses jmh for example.

A few options to introduce this option:
- Introduce a new boolean inc option. Downside this eventually will be removed, likely will touch a few layers in order to pass to `LocalJavaCompiler`
- Reuse the existing inc option `classfileManagerType`, so we only invoke VFS for `TransactionalManagerType`. Downside is `ClassFileManager` likely needs to a new method to optionally wrap an existing FileManager into `ForwardingFM`
- Pass an environment variable w/o touching any interfaces.

The third option seems the cleanest for such an use case. @eed3si9n what do you think? any other ideas?

[1] Compiling a jmh project

```
java.lang.RuntimeException: java.lang.IllegalArgumentException: Not supported: sbt.internal.inc.javac.WriteReportingJavaFileObject@1cfd1875
    at com.sun.tools.javac.main.Main.compile(Main.java:559)
    at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
    at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
    at sbt.internal.inc.javac.LocalJavaCompiler.run(LocalJava.scala:90)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler$$anonfun$compile$1.apply$mcV$sp(AnalyzingJavaCompiler.scala:74)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler$$anonfun$compile$1.apply(AnalyzingJavaCompiler.scala:72)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler$$anonfun$compile$1.apply(AnalyzingJavaCompiler.scala:72)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.timed(AnalyzingJavaCompiler.scala:108)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.compile(AnalyzingJavaCompiler.scala:72)
    at sbt.internal.inc.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply$mcV$sp(MixedAnalyzingCompiler.scala:58)
    at sbt.internal.inc.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:58)
    at sbt.internal.inc.MixedAnalyzingCompiler$$anonfun$compileJava$1$1.apply(MixedAnalyzingCompiler.scala:58)
    at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:73)
    at sbt.internal.inc.MixedAnalyzingCompiler.compileJava$1(MixedAnalyzingCompiler.scala:57)
    at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:63)
    at sbt.internal.inc.IncrementalCompilerImpl$$anonfun$compileInternal$1.apply(IncrementalCompilerImpl.scala:132)
    at sbt.internal.inc.IncrementalCompilerImpl$$anonfun$compileInternal$1.apply(IncrementalCompilerImpl.scala:132)
    at sbt.internal.inc.Incremental$.doCompile(Incremental.scala:95)
    at sbt.internal.inc.Incremental$$anonfun$1$$anonfun$apply$1.apply(Incremental.scala:79)
    at sbt.internal.inc.Incremental$$anonfun$1$$anonfun$apply$1.apply(Incremental.scala:79)
    at sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:70)
    at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:39)
    at sbt.internal.inc.Incremental$$anonfun$1.apply(Incremental.scala:78)
    at sbt.internal.inc.Incremental$$anonfun$1.apply(Incremental.scala:77)
    at sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:122)
    at sbt.internal.inc.Incremental$.compile(Incremental.scala:77)
    at sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:60)
    at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:132)
    at sbt.internal.inc.IncrementalCompilerImpl.incrementalCompile(IncrementalCompilerImpl.scala:87)
    at org.pantsbuild.zinc.Compiler.compile(Compiler.scala:207)
    at org.pantsbuild.zinc.Main$.run(Main.scala:97)
    at org.pantsbuild.zinc.Main$.main(Main.scala:16)
    at org.pantsbuild.zinc.Main.main(Main.scala)
Caused by: java.lang.IllegalArgumentException: Not supported: sbt.internal.inc.javac.WriteReportingJavaFileObject@1cfd1875
    at com.sun.tools.javac.file.JavacFileManager.isSameFile(JavacFileManager.java:647)
    at javax.tools.ForwardingJavaFileManager.isSameFile(ForwardingJavaFileManager.java:90)
    at com.sun.tools.javac.api.ClientCodeWrapper$WrappedJavaFileManager.isSameFile(ClientCodeWrapper.java:257)
    at com.sun.tools.javac.processing.JavacFiler.checkFileReopening(JavacFiler.java:532)
    at com.sun.tools.javac.processing.JavacFiler.getResource(JavacFiler.java:483)
    at org.openjdk.jmh.generators.annotations.APGeneratorDestinaton.getResource(APGeneratorDestinaton.java:56)
    at org.openjdk.jmh.generators.core.BenchmarkGenerator.readBenchmarkList(BenchmarkGenerator.java:194)
    at org.openjdk.jmh.generators.core.BenchmarkGenerator.complete(BenchmarkGenerator.java:135)
    at org.openjdk.jmh.generators.BenchmarkProcessor.process(BenchmarkProcessor.java:60)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:794)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$200(JavacProcessingEnvironment.java:91)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.runContributingProcs(JavacProcessingEnvironment.java:627)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1033)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1198)
    at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
    at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
    at com.sun.tools.javac.main.Main.compile(Main.java:523)
    ... 32 more
[error] java.lang.IllegalArgumentException: Not supported: sbt.internal.inc.javac.WriteReportingJavaFileObject@1cfd1875
```

--- a/build.sbt
+++ b/build.sbt
@@ -64,7 +64,7 @@ def commonSettings: Seq[Setting[_]] = Seq(
     new URL("https://dl.bintray.com/sbt/ivy-snapshots/")
   )(Resolver.ivyStylePatterns),
   // concurrentRestrictions in Global += Util.testExclusiveRestriction,
-  testOptions += Tests.Argument(TestFrameworks.ScalaCheck, "-w", "1"),
+  testOptions += Tests.Argument(TestFrameworks.ScalaCheck, "-w", "1", "-verbosity", "2"),
   javacOptions in compile ++= Seq("-Xlint", "-Xlint:-serial"),
   crossScalaVersions := Seq(scala212),
   publishArtifact in Test := false,

--- a/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/LocalJava.scala
+++ b/internal/zinc-compile-core/src/main/scala/sbt/internal/inc/javac/LocalJava.scala
@@ -255,11 +255,6 @@ final class LocalJavaCompiler(compiler: javax.tools.JavaCompiler) extends XJavaC
  *
  * @param fileManager A manager for Java files.
  * @param classFileManager The instance that manages generated class files.
- *
- * @note `getFileForOutput` used by the annotation process for writing
- *       resources cannot be overridden because of a Javac SDK check.
- *       JDK8 has a hard-coded check against it that impedes wrapping
- *       `RegularFileObject` with other instances, e.g. `ForwardingFileObject`.
  */
 final class WriteReportingFileManager(
     fileManager: JavaFileManager,
@@ -274,6 +269,17 @@ final class WriteReportingFileManager(
     val output = super.getJavaFileForOutput(location, className, kind, sibling)
     new WriteReportingJavaFileObject(output, classFileManager)
   }
+
+  // Fixes #185, #684
+  override def isSameFile(a: FileObject, b: FileObject): Boolean = {
+    def unwrap(fo: FileObject): FileObject = {
+      fo match {
+        case wrapper: WriteReportingJavaFileObject => wrapper.javaFileObject
+        case notWrapped                            => notWrapped
+      }
+    }
+    super.isSameFile(unwrap(a), unwrap(b))
+  }
 }
 
 /**
@@ -283,7 +289,7 @@ final class WriteReportingFileManager(
  * @param classFileManager The instance that manages generated class files.
  */
 final class WriteReportingJavaFileObject(
-    javaFileObject: JavaFileObject,
+    val javaFileObject: JavaFileObject,
     var classFileManager: ClassFileManager
 ) extends ForwardingJavaFileObject[JavaFileObject](javaFileObject) {
   override def openWriter(): Writer = {

